src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,304 @@
+}
+
+void Qgs3DMapToolMeasureLine::onTerrainPicked( Qt3DRender::QPickEvent *event )
+{
+  qInfo() << "onTerrainPicked";
+  if ( event->button() == Qt3DRender::QPickEvent::LeftButton )
+  {
+    if ( mDone )
+    {
+      mDialog->restart();
+    }
+
+    const QVector3D worldIntersection = event->worldIntersection();
+    QgsVector3D mapCoords = Qgs3DUtils::worldToMapCoordinates( QgsVector3D( worldIntersection.x(),
+                            worldIntersection.y(),
+                            worldIntersection.z() ), mCanvas->map()->origin() );
+    qInfo() << "Coord (onTerrainPicked): " << mapCoords.x() << " " << mapCoords.y() << " " << mapCoords.z();
+
+    mDone = false;
+    addPoint( QgsPoint( mapCoords.x(), mapCoords.y(), mapCoords.z() ) );
+    mDialog->show();
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::RightButton )
+  {
+    // Finish measurement
+    qInfo() << "Finish measurement";
+    mDone = true;
+  }
+  else if ( event->button() == Qt3DRender::QPickEvent::MiddleButton )
+  {
+    qInfo() << "Undo - middle button";
+    undo();
+  }
+
+
+
+}
+
+
+void Qgs3DMapToolMeasureLine::onTerrainEntityChanged()
+{
+  qInfo() << "onTerrainEntityChanged";
+  // no need to disconnect from the previous entity: it has been destroyed
+  // start listening to the new terrain entity
+  if ( QgsTerrainEntity *terrainEntity = mCanvas->scene()->terrainEntity() )
+  {
+    connect( terrainEntity->terrainPicker(), &Qt3DRender::QObjectPicker::clicked, this, &Qgs3DMapToolMeasureLine::onTerrainPicked );
+  }
+}
+
+void Qgs3DMapToolMeasureLine::setMeasurementLayerRenderer( QgsVectorLayer *layer )
+{
+  layer->setRenderer3D( mLineSymbolRenderer );
+  mLineSymbolRenderer->setLayer( layer );
+}
+
+
+void Qgs3DMapToolMeasureLine::addPoint( const QgsPoint &point )
+{
+  // don't add points with the same coordinates
+  if ( !mPoints.isEmpty() && mPoints.last() == point )
+  {
+    return;
+  }
+
+  QgsPoint addedPoint( point );
+  qInfo() << "Is 3D point? " << addedPoint.is3D();
+
+  // Append point
+  mPoints.append( addedPoint );
+
+  mMeasurementLine->addVertex( addedPoint );
+
+  mMeasurementLayer->startEditing();
+  QgsGeometry *newMeasurementLine = new QgsGeometry( mMeasurementLine );
+  qInfo() << "Current line: " << newMeasurementLine->asWkt();
+  mMeasurementLayer->changeGeometry( mMeasurementFeature->id(), *newMeasurementLine );
+  mMeasurementLayer->commitChanges();
+
+  mDialog->addPoint();
+
+  mCanvas->map()->setRenderers( QList<QgsAbstract3DRenderer *>() << mMeasurementLayer->renderer3D()->clone() );
+}
+
+void Qgs3DMapToolMeasureLine::restart()
+{
+  mPoints.clear();
+
+  mDone = true;
+  mMeasurementLine->clear();
+
+  mMeasurementLayer->startEditing();
+  QgsGeometry *newMeasurementLine = new QgsGeometry( mMeasurementLine );
+  qInfo() << "Current line: " << newMeasurementLine->asWkt();
+  mMeasurementLayer->changeGeometry( mMeasurementFeature->id(), *newMeasurementLine );
+  mMeasurementLayer->commitChanges();
```

Comment:
It would be good to have a single function e.g. updateMeasurementLayer() that would change geometry and commit changes based on mPoints - and multiple functions can use it do de-duplicate code.

Reply:
Done